devtools/x/src/cargo.rs

```diff
@@ -437,3 +392,153 @@ impl<'a> CargoCommand<'a> {
+            if self.changed_since.is_some() {
+                exclusive.push("--changed-since");
+            }
+            if !self.package.is_empty() {
+                exclusive.push("--package");
+            }
+            if self.workspace {
+                exclusive.push("--workspace");
+            }
+
+            if exclusive.len() > 1 {
+                let err_msg = exclusive.join(", ");
+                return Err(anyhow!("can only specify one of {}", err_msg));
+            }
+        }
+
+        if self.workspace {
+            Ok(CargoPackages::workspace())
+        } else if !self.package.is_empty() {
+            Ok(CargoPackages::includes(
+                self.package.iter().map(|s| s.as_str()),
+            ))
+        } else if let Some(base) = &self.changed_since {
+            let affected_set = changed_since_impl(&xctx, &base)?;
+
+            Ok(CargoPackages::includes(
+                affected_set
+                    .packages(DependencyDirection::Forward)
+                    .map(|package| package.name()),
+            ))
+        } else {
+            CargoPackages::default_cwd(xctx)
+        }
+    }
+}
+
+/// Package selector for Cargo commands.
+#[derive(Clone, Debug)]
+pub struct CargoPackages<'a> {
+    includes: CargoPackageInclude<'a>,
+    excludes: BTreeSet<&'a str>,
+}
+
+impl<'a> CargoPackages<'a> {
+    /// Returns a new `CargoPackages` that selects all packages in this workspace.
+    pub fn workspace() -> Self {
+        Self {
+            includes: CargoPackageInclude::Workspace,
+            excludes: BTreeSet::new(),
+        }
+    }
+
+    /// Returns a new `CargoPackages` that selects the default set of packages for the current
+    /// working directory. This may either be the entire workspace or a local package.
+    pub fn default_cwd(xctx: &XContext) -> Result<Self> {
+        let includes = if project_is_root(&xctx.config().cargo_config())? {
+            CargoPackageInclude::Workspace
+        } else {
+            CargoPackageInclude::LocalPackage
+        };
+        Ok(Self {
+            includes,
+            excludes: BTreeSet::new(),
+        })
+    }
+
+    /// Returns a new `CargoPackages` that selects the specified packages.
+    pub fn includes(package_names: impl IntoIterator<Item = &'a str>) -> Self {
+        Self {
+            includes: CargoPackageInclude::Includes(package_names.into_iter().collect()),
+            excludes: BTreeSet::new(),
+        }
+    }
+
+    /// Adds excludes for this `CargoPackages`.
+    ///
+    /// The excludes are currently ignored if the local package is built.
+    pub fn add_excludes(&mut self, exclude_names: impl IntoIterator<Item = &'a str>) -> &mut Self {
+        self.excludes.extend(exclude_names);
+        self
+    }
+
+    // ---
+    // Helper methods
+    // ---
+
+    fn is_empty(&self) -> bool {
+        match &self.includes {
+            CargoPackageInclude::Workspace | CargoPackageInclude::LocalPackage => false,
```

Comment:
I can see for testing purposes why this is has nothing do with what's on disk, not sure if it's a good thing or not yet.

Reply:
Hmm, could you elaborate? I'm not sure what you mean.

Reply:
TODO: change name of method, add logging for the case where we don't try and build anything.

Reply:
Done.